src/components/Swap/Summary/index.tsx

```diff
@@ -27,59 +30,176 @@ const PriceImpactText = styled.span`
   color: ${({ theme }) => theme.error};
 `
 
+const PermitTooltipBody = styled.div`
+  max-width: 220px;
+`
+
+function PermitTooltipText({ text, content }: { text: ReactNode; content: ReactNode }) {
+  return (
+    <TooltipText placement="bottom" offset={10} text={text}>
+      <PermitTooltipBody>
+        <ThemedText.Caption>{content}</ThemedText.Caption>
+      </PermitTooltipBody>
+    </TooltipText>
+  )
+}
+
+function getAllowanceFailedAction(isApproved: boolean, retry: () => void): Action {
+  return {
+    message: isApproved ? (
+      <PermitTooltipText
+        text={t`Token approval failed`}
+        content={t`A signature is needed to trade this token on the Uniswap protocol. For security, signatures expire after 30 days.`}
+      />
+    ) : (
+      <PermitTooltipText
+        text={t`Permit approval failed`}
+        content={t`Permit2 allows safe sharing and management of token approvals across different smart contracts.`}
+      />
+    ),
+    onClick: retry,
+    color: 'warning',
+    children: <Trans>Try again</Trans>,
+  }
+}
+
+function getAllowancePendingAction(isApproved: boolean, cancel: () => void): Action {
+  return {
+    message: isApproved ? (
+      <PermitTooltipText
+        text={t`Approve token for trading`}
+        content={t`Gives you the ability to trade this token on the Uniswap protocol. For security, this will expire in 30 days.`}
+      />
+    ) : (
+      <PermitTooltipText
+        text={t`Approve permit`}
+        content={t`Permit2 allows safe sharing and management of token approvals across different smart contracts.`}
+      />
+    ),
+    icon: Spinner,
+    onClick: cancel,
+    children: <Trans>Cancel</Trans>,
+  }
+}
+
+enum ReviewState {
+  REVIEWING,
+  ALLOWANCE_PENDING,
+  ALLOWANCE_FAILED,
+  TRADE_CHANGED,
+  PENDING_SWAP,
```

Comment:
nit: `SWAP_PENDING` seems more consistent with the other enums 😅

Reply:
true, will do